### PR TITLE
Feature: Repo indexing + repo state machine to manage concurrent requests correctly

### PR DESCRIPTION
**Pull request by timotheeguerin**
This adds the following features: 
- Keep an index of repos which is resilient to restart to keep track of most used repos(and potentially more data later on)
- Handle concurrency of accessing a repo correctly between open/read/fetch/delete actions
- Recover from a repo cloned in a bad state(Will delete and clone again)

### CODE REVIEW

File: src/services/repo-index/repo-index.service.ts [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+export interface RepoReference {
+  readonly path: string;
+  readonly lastUse: number;
+  readonly lastFetch?: number;
+}
+
+const FETCH_CACHE_EXPIRY = 30_000; // 30s;
+
+@Injectable()
+export class RepoIndexService {
```
Comment (vladbarosan) [inline src/services/repo-index/repo-index.service.ts]: >RepoIndexService [](start = 13, length = 16)

comments on what it does

File: src/services/repo-index/repo-index.service.ts [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+  private repository: Repository<RepoReferenceRecord>;
+
+  constructor(connection: Connection) {
+    this.repository = connection.getRepository(RepoReferenceRecord);
+    this.init().catch(e => {
+      this.logger.error("Failed to load data from database", e);
+    });
+  }
+
+  public async init() {
```
Comment (vladbarosan) [inline src/services/repo-index/repo-index.service.ts]: >init [](start = 15, length = 4)

any reason to have this public ? seems that the constructor calls it

File: src/core/logger/logger.ts [hunk 3]
```diff
@@ -7,10 +7,14 @@ export interface LogMetadata {
   [key: string]: any;
 }
 
+type Class = new (...args: any[]) => any;
```
Comment (vladbarosan) [inline src/core/logger/logger.ts]: >any [](start = 37, length = 3)

can it be unknown ?

File: src/services/repo-index/repo-index.service.ts (hunk 1 above)
Comment (vladbarosan) [reply src/services/repo-index/repo-index.service.ts]: same for all the code

---
In reply to: [300110784](https://github.com/Azure/git-rest-api/pull/51#discussion_r300110784) [](ancestors = 300110784)

File: src/services/repo-index/repo-index.service.ts (hunk 2 above)
Comment (timotheeguerin) [reply src/services/repo-index/repo-index.service.ts]: this one actually no

File: src/controllers/branches/branches.controller.e2e.ts [hunk 4]
```diff
@@ -3,7 +3,7 @@ import { delay } from "../../utils";
 
 describe("Test branch controller", () => {
   it("doesn't conflict when getting the same repo twice at the same time", async () => {
-    await delay(100);
+    await delay(2000);
     await deleteLocalRepo(UNENCODED_TEST_REPO);
```
Comment (vladbarosan) [inline src/controllers/branches/branches.controller.e2e.ts]: what does this delay is meant to do ? (document pleeaase)

File: src/core/logger/winston-logger.ts [hunk 5]
```diff
@@ -59,6 +59,10 @@ if (config.nodeEnv === "development") {
   );
 }
 
+if (config.nodeEnv === "test") {
+  consoleTransport.silent = true;
+}
```
Comment (vladbarosan) [inline src/core/logger/winston-logger.ts]: why silent in test ?

**Comment by vladbarosan** [comment]
> if (config.nodeEnv === "development") {

this one here is dev,

---
Refers to: src/core/logger/winston-logger.ts:46 in 4d48ac5. [](commit_id = 4d48ac5e6f4083655047d72ec5ad710f944907cf, deletion_comment = False)

File: src/services/fs/fs.service.ts [hunk 6]
```diff
@@ -1,5 +1,9 @@
 import { Injectable } from "@nestjs/common";
 import fs from "fs";
+import rimraf from "rimraf";
+import { promisify } from "util";
+
+const rm = promisify(rimraf);
```
Comment (vladbarosan) [inline src/services/fs/fs.service.ts]: >rimraf [](start = 21, length = 6)

this one seems to be done as a dev dependency

Comment (vladbarosan) [reply src/services/fs/fs.service.ts]: also you dont even need to define this class, you can use fs-xtra. [https://www.npmjs.com/package/fs-extra](https://www.npmjs.com/package/fs-extra)

---
In reply to: [300143207](https://github.com/Azure/git-rest-api/pull/51#discussion_r300143207) [](ancestors = 300143207)

File: src/services/repo/mutex/mutex.ts [hunk 7]
```diff
@@ -0,0 +1,68 @@
...
+export interface Lock {
+  readonly id: string;
+  readonly release: () => void;
+}
+
+export interface LockOptions {
+  exclusive: boolean;
+}
+
+export class Mutex {
```
Comment (vladbarosan) [inline src/services/repo/mutex/mutex.ts]: >Mutex [](start = 13, length = 5)

no comments... please comment all the code added

**Pull request merged**